File: LeetCode/2946/main.cpp
class Solution {
public:
    bool areSimilar(vector<vector<int>>& mat, int k) {
        int m = mat.size();
        int n = mat[0].size();

        // for(int i = 0; i < m; i++) {
        //     for(int j = 0; j < n; j++) {
        //         if(i % 2 == 1) {
        //             // left shift
        //             // actually (j-k+n) % n but we do ((j-k)%n + n) % n to avoid overflow
        //             if(mat[i][j] != mat[i][((j-k)%n+n)%n]) {
        //                 return false;
        //             }
        //         } else {
        //             // right shift
        //             if(mat[i][j] != mat[i][(j+k)%n]) {
        //                 return false;
        //             }
        //         }
        //     }
        // }

        if(k % n == 0) {
            return true;
        }
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if(mat[i][j] != mat[i][(j+k) % n]) {
                    return false;
                }
            }
        }

        return true;

    }
};

File: LeetCode/2946/main.py
class Solution:
    def areSimilar(self, mat: List[List[int]], k: int) -> bool:

        # def rshift(arr, k) :
        #     n = len(arr)
        #     k %= n
        #     return arr[-k:] + arr[:n-k]

        # def lshift(arr, k) :
        #     n = len(arr)
        #     k %= n
        #     return arr[k:] + arr[:k]

        m, n = len(mat), len(mat[0])

        for i in range(m):
            for j in range(n):
                if (i % 2 == 0):
                    # if (j-k) < 0, then just add n else leave it as it is
                    if (mat[i][j] != mat[i][(j-k+n) % n]):
                        return False
                else:
                    if (mat[i][j] != mat[i][(j+k) % n]):
                        return False

        return True
